BJS2-26098

Silencemess1ah opened the pull request.

Silencemess1ah committed:
BJS2-26098

badasscoding commented on src/main/java/faang/school/postservice/controller/LikeController.java:
```diff
@@ -0,0 +1,30 @@
+
+    private final LikeService likeService;
+
+    @GetMapping("/post/like/{id}")
```
1. Название путей не соответствует RESTful, у тебя как-будто по id ищется лайк
2. Плюс обычно стараются использовать существительные во множественном числе


Возможен такой вариант:

```
@GetMapping("/posts/{id}/likes")
@ResponseBody
public List<UserDto> getAllUsersByPost(@PathVariable Long id) {
    return likeService.getAllUsersByPostId(id);
}

@GetMapping("/comments/{id}/likes")
@ResponseBody
public List<UserDto> getAllUsersByComment(@PathVariable Long id) {
    return likeService.getAllUsersByCommentId(id);
}
```

Silencemess1ah replied:
Согласен, поправлю, учту!

mighty-mhsl replied:
Согласен, надо поправить

badasscoding commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+
+    public List<UserDto> getAllUsersByPostId(long id) {
+        List<Long> userIds = getUsersIdsByPostId(id);
+        if (userIds.size() > 100) {
```
кажется что можно убрать if и всё через один метод сделать (у тебя один раз вызовется цикл и int batchEnd = Math.min(i + 100, userIds.size()); корректно отработает

mighty-mhsl replied:
Что такое 100? Это выглядит, как magic number. Должна быть константа с описательным именем

badasscoding commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+
+    public List<UserDto> getAllUsersByCommentId(long id) {
+        List<Long> userIds = getUserIdsByCommentId(id);
+        if (userIds.size() > 100) {
```
и тут так же

badasscoding commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+        List<UserDto> usersLiked = new ArrayList<>();
+        for (int i = 0; i < batchSize; i += 100) {
+            int batchEnd = Math.min(i + 100, userIds.size());
+            usersLiked.addAll(userServiceClient.getUsersByIds(userIds.subList(i, batchEnd)));
```
тут бы не в одну строчку сделать для лучшей читаемости

badasscoding commented on src/test/java/faang/school/postservice/controller/LikeControllerTest.java:
```diff
@@ -0,0 +1,54 @@
+
+@ExtendWith(MockitoExtension.class)
+public class LikeControllerTest {
+    @InjectMocks
```
пустой строки не хватает

badasscoding commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+    private List<UserDto> usersLiked;
+    private List<Like> likes;
+
+
```
лишняя пустая строка

Silencemess1ah committed:
Fixes

mighty-mhsl commented on src/main/java/faang/school/postservice/controller/LikeController.java:
```diff
@@ -0,0 +1,30 @@
+    private final LikeService likeService;
+
+    @GetMapping("/post/like/{id}")
+    @ResponseBody
```
Если у тебя висит аннотация @RestController на классе, то @ResponseBody добавляется на эндпоинты автоматом. Так что можно убрать

mighty-mhsl commented on src/main/java/faang/school/postservice/controller/LikeController.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @GetMapping("/comment/like/{id}")
+    @ResponseBody
```
И тут

mighty-mhsl commented on src/main/java/faang/school/postservice/controller/LikeController.java:
```diff
@@ -0,0 +1,30 @@
+        return likeService.getAllUsersByPostId(id);
+    }
+
+    @GetMapping("/comment/like/{id}")
```
И тут путь стоит поправить
/comment/{id}/like

mighty-mhsl commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+        return likes.stream()
+                .map(Like::getUserId)
+                .toList();
+    }
```
Эти два метода выглядят, как абсолютно идентичные. В них только вызов в БД отличается. Нам точно нужно столько дублирования?

mighty-mhsl commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+        if (userIds.size() > 100) {
+            return getUserDtoListByIdsBatched(userIds);
+        }
+        return userServiceClient.getUsersByIds(userIds);
```
Здесь тоже дважды пишем одинаковые методы, хотя у нас только первая строчка в них отличается. Почему дублируем?

mighty-mhsl commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+        long batchSize = userIds.size() / 100 + 1;
+        List<UserDto> usersLiked = new ArrayList<>();
+        for (int i = 0; i < batchSize; i += 100) {
+            int batchEnd = Math.min(i + 100, userIds.size());
```
Повсюду 100. Что такое 100?

mighty-mhsl commented on src/main/java/faang/school/postservice/service/LikeService.java:
```diff
@@ -0,0 +1,59 @@
+                .toList();
+    }
+
+    private List<UserDto> getUserDtoListByIdsBatched(List<Long> userIds) {
```
getUsersBatched можно назвать - будет короче. Слово Dto можно не использовать в названии метода, потому что мы и так видим, что возвращается дто - это детали кода, а не логики

mighty-mhsl commented on src/test/java/faang/school/postservice/controller/LikeControllerTest.java:
```diff
@@ -0,0 +1,54 @@
+        public void whenServiceCalledGetAllUsersByCommentIdThenReturnList() {
+            when(likeService.getAllUsersByCommentId(COMMENT_ID_ONE)).thenReturn(usersLiked);
+            likeController.getAllUsersByComment(COMMENT_ID_ONE);
+            verify(likeService).getAllUsersByCommentId(COMMENT_ID_ONE);
```
А почему мы не делаем assrt того, что нам usersLiked вернулись из контроллера тогда, раз сервис их возвращает? Просто проверили, что вызываем этот метод, но не проверили, что он нам вернул то, что мы ожидаем

mighty-mhsl commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+        public void whenPostIdPassedThenReturnUserDtoWhoLikedPostList() {
+            when(likeService.getAllUsersByPostId(POST_ID_ONE)).thenReturn(usersLiked);
+            likeService.getAllUsersByPostId(POST_ID_ONE);
+            assertEquals(usersLiked.size(), userIds.size());
```
А наполнение списков не будем проверять? Ведь размер может совпадать, но элементы могут быть ошибочными

mighty-mhsl commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+        public void whenCommentIdPassedThenReturnUserDtoWhoLikedCommentList() {
+            when(likeService.getAllUsersByCommentId(COMMENT_ID_ONE)).thenReturn(usersLiked);
+            likeService.getAllUsersByCommentId(COMMENT_ID_ONE);
+            assertEquals(usersLiked.size(), userIds.size());
```
Наполнение списка проверить

mighty-mhsl commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+        public void whenCommentIdPassedAndListIsMoreThanOneHundredThenBatchesMade() {
+            when(likeRepository.findByCommentId(COMMENT_ID_ONE)).thenReturn(likes);
+
+            when(likeService.getAllUsersByCommentId(COMMENT_ID_ONE)).thenReturn(usersLiked);
```
Как мы можем мокнуть метод, который сами и тестируем тут? Зачем его мокать, если нам НУЖНО, чтобы он по-настоящему отработал

mighty-mhsl commented on src/test/java/faang/school/postservice/service/LikeServiceTest.java:
```diff
@@ -0,0 +1,106 @@
+        public void whenPostIdPassedAndListIsMoreThanOneHundredThenBatchesMade() {
+            when(likeRepository.findByPostId(POST_ID_ONE)).thenReturn(likes);
+
+            when(likeService.getAllUsersByPostId(POST_ID_ONE)).thenReturn(usersLiked);
```
И здесь
Что мы тестируем? Конечно, getAllUsersByPostId вернем мне usersLiked, если я в этом же тесте сам его попросил это возвращать. Тогда что мы тестируем? Сам тест? Или логику?
Выглядит так, что мы тестируем сам тест.

Silencemess1ah committed:
Deleted requestBody annotation
fixed urls
added constant for batch size in service
fixed tests assert and some cases
fixed service duplication

Silencemess1ah committed:
imports delete
